File: src/server/game/config/deck_types.go
package config

import "sync"

type DeckType struct {
	Name            string   `json:"name"`
	CardLimitLower  int      `json:"card_limit_lower"`
	CardLimitUpper  int      `json:"card_limit_upper"`
	Public          bool     `json:"public"`
	CopiesLimit     int      `json:"copies_limit"`
	SharedLimitWith []string `json:"shared_limit_with"`
}

type DeckTypesConfig struct {
	MainDeck  string     `json:"main_deck"`
	DeckTypes []DeckType `json:"decks_types"`
}

var (
	deckTypesConfig    sync.Once
	DeckTypesConfigVar DeckTypesConfig
)

// GetDeckTypesConfig allows to get the game configuration linked to deck types globally.
func GetDeckTypesConfig() DeckTypesConfig {
	return DeckTypesConfigVar
}

File: src/game/decks/deck_types.lua
local DeckType = {}
DeckType.__index = DeckType

function DeckType:new(name, card_limit_lower, card_limit_upper, public, copies_limit, shared_limit_with)
    local deckType = setmetatable({}, DeckType)

    deckType.name = name
    deckType.card_limit_lower = card_limit_lower
    deckType.card_limit_upper = card_limit_upper
    deckType.public = public or false
    deckType.copies_limit = copies_limit or 1
    deckType.shared_limit_with = shared_limit_with or {}

    return deckType
end

return DeckType
